Fix Custom Resource update bugs and external registry URLs not taken in account

davidfestal opened the pull request:
This PR fixes a number of weirds behaviors / hard bugs happening when the `CheCluster` Custom Resource is updated.

It fixes issue https://github.com/eclipse/che/issues/14111

nickboldt commented:
Not sure I'm qualified to review this one.

davidfestal commented:
@l0rd Do you plan to review it finally ?

rhopp commented:
@davidfestal I've tried this and seems to be working reasonably well with some exceptions (I'm not sure all are related to this PR)
1) After first CR edit, I'm getting `time="2019-08-06T14:44:00Z" level=info msg="Updating deployment che with new memory settings. Request: 512Mi, limit: 1Gi" ` every few seconds - the `che` deployment is edited every few seconds.
2) Changing `externalDevfileRegistry` to `true` from `false` does use external registry, but the local registry is not undeployed - I guess it should be? This is not blocking 7.0 anyway.
3) Changing `externalDevfileRegistry` to `false` from `true`, but leaving `devfileRegistryUrl` not empty does nothing. Only deleting `devfileRegistryUrl` edits the CM and redeploys che-server.

I didn't have enough time to try all possible scenarios, but I think at least 1) and 3) should be addressed.

Sidenote - This is the type of PR which would be great, if it was covered by tests... We can create a separate issue for this.

nickboldt commented on pkg/controller/che/che_controller.go:
```diff
@@ -470,20 +477,19 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
 		}
 		if cheFlavor != "codeready" {
-			pluginRegistryURL += "/v3"
+			guessedPluginRegistryURL += "/v3"
```
Should this be externalized as a property somehow? like a default registry version variable, rather than encoded this way?

davidfestal replied:
this could probably be externalized, but since it was already like that before this PR, I would let this for a next PR, no ?

nickboldt replied:
yeah, save it for later refactoring. I'll assume you'll open an issue for externalizing params and refactoring "keycloak" to "identityProvider" and target it for Che 7.2 or later.

nickboldt commented on pkg/controller/che/che_controller.go:
```diff
@@ -470,20 +477,19 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
-			instance, _ = r.GetCR(request)
-			return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 1}, err
+		if pluginRegistryURL == "" {
+			pluginRegistryURL = guessedPluginRegistryURL
```
Does it matter that you're no longer throwing a status report here?

davidfestal replied:
The status report is thrown in the same way, but just some lines later, where this block has indeed been moved. because we have to update the registry URL on the status in both cases (when using embedded registry or when an external registry URL is specified)

nickboldt commented on pkg/controller/che/che_controller.go:
```diff
@@ -499,18 +505,24 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
 	}
+	if pluginRegistryURL != instance.Status.PluginRegistryURL {
+		instance.Status.PluginRegistryURL = pluginRegistryURL
+		if err := r.UpdateCheCRStatus(instance, "status: Plugin Registry URL", pluginRegistryURL); err != nil {
```
Ah, I see the block deleted from line 491 has moved down here. So this is the fallback value if the guess doesn't work?

davidfestal replied:
more precisely, if an external URL is provided we use it (even if the embedded registry is deployed => a guessed value is available, but finally not used).

nickboldt commented on pkg/controller/che/che_controller.go:
```diff
@@ -448,7 +448,14 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
+			actualMemLimit := effectiveDeployment.Spec.Template.Spec.Containers[0].Resources.Limits[corev1.ResourceMemory]
+			limitStr := actualMemLimit.String()
+			requestStr := actualMemRequest.String()
+			if effectiveDeployment.Spec.Template.Spec.Containers[0].Image != registryImage ||
```
why do we need these extra checks here?

davidfestal replied:
To identify situations when caracteristics of the current deployment are not the same as those required by the custom resource. In such a situation, we should redeploy => restart the POD.

nickboldt commented on pkg/controller/che/che_controller.go:
```diff
@@ -720,32 +729,32 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
 		cheImageTag = util.GetValue(instance.Spec.Server.CheImageTag, deploy.DefaultCodeReadyServerImageTag)
 	}
-	cheDeployment, err := deploy.NewCheDeployment(instance, cheImageRepo, cheImageTag, cmResourceVersion, isOpenShift)
+	cheDeploymentToCreate, err := deploy.NewCheDeployment(instance, cheImageRepo, cheImageTag, cmResourceVersion, isOpenShift)
```
you have variables called `newKeycloakDeployment` and `cheDeploymentToCreate` ... assuming these are both new things to be created, should the names be more consistent? eg., both prefix with `new` or suffix with `ToCreate` ?

Also any chance we can call it "IdentityProvider" instead of "Keycloak" ? :) Or is that a bad refactoring to do piecemeal, which we should tackle in Che 8.0 ?

davidfestal replied:
I already tried to clean naming to make the code easier to read / understand in the most critical parts. But I might have forgotten some of left some inconsistencies. Is it a blocker for you ?

davidfestal replied:
As for the `Keycloak` to `ìdentityProvider` change, I'd prefer doing that in a next PR, probably in the context of the wider refactoring that will certainly be required after GA.

nickboldt replied:
not a blocker. +1 for wider refactoring in another PR.

nickboldt commented on pkg/controller/che/che_controller.go:
```diff
@@ -861,14 +870,18 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
+	if desiredRequest != requestStr ||
+		desiredLimit != limitStr ||
+		string(effectiveCheDeployment.Spec.Template.Spec.Containers[0].ImagePullPolicy) != desiredImagePullPolicy ||
+		(r.GetDeploymentEnv(effectiveCheDeployment, "CHE_SELF__SIGNED__CERT") != "") != instance.Spec.Server.SelfSignedCert {
```
these last 2 checks are new. Why do we need them?

davidfestal replied:
Because these values are used in the Che deployment, so when they are changed, the deployment should be updated and the underlying POD restarted. It was a bug that they were not.

nickboldt replied:
+1

davidfestal commented:
> After first CR edit, I'm getting time="2019-08-06T14:44:00Z" level=info msg="Updating deployment che with new memory settings. Request: 512Mi, limit: 1Gi" every few seconds - the che deployment is edited every few seconds.

How many times is the deployment edited, and how long is it regularly edited ? Does it stop after some time ?

davidfestal commented:
> Changing externalDevfileRegistry to false from true, but leaving devfileRegistryUrl not empty does nothing. Only deleting devfileRegistryUrl edits the CM and redeploys che-server.

That doesn't seem a problem to me: as soon as you have `devfileRegistryURL` set to a non-empty value, the value is used in the Che config map, even if `externalDevfileRegistry` is `false`.
We don't enfore consistency between those 2 properties, and the URL property setting has the priority.

davidfestal commented:
> Changing externalDevfileRegistry to true from false does use external registry, but the local registry is not undeployed - I guess it should be?

Yes, it's a known limitation.

> This is not blocking 7.0 anyway

You're right, since implementation would have required more complex and impacting changes, I prefered letting this for after GA.

rhopp commented:
@davidfestal Here's what I did:
* `chectl server:start -a operator -p openshift -b crw.codeready.com -n rhopp-test --che-operator-image=quay.io/rhopp/che-operator:cr-update-bugs`  (che-operator:cr-update-bugs is image I built from this PR)
* I waited until the default deployment finishied successfully
* Then I changed two values in checluster CR:
* * `spec.server.tlsSupport: false` to `spec.server.tlsSupport: true`
* * `spec.server.selfSignedCert: false` to `spec.server.selfSignedCert: true`
* Then I let the operator do it's business.
* Once operator finished with creating routes, redeploying che etc (I don't exactly know what everything is it doing) - At this point I have usable "secure" che.

But when I look at che-operator logs, I can see, that every 5 seconds, it is doing something wich che deployment - this seems like a bug which needs to be fixed before merging.

**che-operator logs**: https://pastebin.com/jD31yHiQ

davidfestal commented:
@rhopp I fixed the bug you reported with the operator constantly deploying the Che server after a first CR update to TLS mode. I also fixed a similar issue with memory request / limit fields for the `che` and registry deployments.

I also rebased the PR on the last master, but the related commits are 5cee451, 9c32120 and 40c01b9.

@nickboldt FTR

rhopp commented:
I've tried `quay.io/dfestal/che-operator:fix-update-bugs` - LGTM.
